lib/wibox/layout/ratio.lua

```diff
@@ -0,0 +1,339 @@
+    local index = self:index(widget)
+
+    self:set_ratio(index, percent)
+end
+
+--- Update all widgets to match a set of a ratio
+-- @tparam number index The widget to change or its index
+-- @tparam number before The sum of the ratio before the widget
+-- @tparam number itself The ratio for "widget"
+-- @tparam number after The sum of the ratio after the widget
+function ratio:ajust_ratio(index, before, itself, after)
+    if not self.widgets[index] or not before or not itself or not after then
+        return
+    end
+
+    -- Compute the before and after offset to be applied to each widgets
+    local before_count, after_count = index-1, #self.widgets - index
+
+    local b, a = gen_sum(self, 1, index-1), gen_sum(self, index+1)
+
+    local db, da = (before - b)/before_count, (after - a)/after_count
+
+    -- Apply the new ratio
+    self._ratios[index] = itself
+
+    -- Equality split the delta among widgets before and after
+    for i = 1, index -1 do
+        self._ratios[i] = self._ratios[i] + db
+    end
+    for i = index+1, #self.widgets do
+        self._ratios[i] = self._ratios[i] + da
+    end
+
+    normalize(self)
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Update all widgets to match a set of a ratio
+-- @param widget The widget to ajust
+-- @tparam number before The sum of the ratio before the widget
+-- @tparam number itself The ratio for "widget"
+-- @tparam number after The sum of the ratio after the widget
+function ratio:ajust_widget_ratio(widget, before, itself, after)
+    local index = self:index(widget)
+    self:ajust_ratio(index, before, itself, after)
+end
+
+--- Add some widgets to the given fixed layout
+-- @tparam widget ... Widgets that should be added (must at least be one)
+function ratio:add(...)
+    fixed.add(self,...)
+
+    normalize(self)
+end
+
+--- Remove one or more widgets from the layout
+-- @tparam widget ... Widgets that should be removed (must at least be one)
+-- The last parameter can be a boolean, forcing a `recursive` search of the
+-- widget(s) to remove.
+function ratio:remove_widgets(...)
+    -- Clear the cache
+    local count = #self._ratios - #({...})
+    for i=count+1, #self._ratios do
+        self._ratios[i] = nil
+    end
+
+    fixed.remove_widgets(self,...)
+
+    normalize(self)
+end
+
+--- Remove a widget from the layout
+-- @tparam number index The widget index to remove
+-- @treturn boolean index If the operation is successful
+function ratio:remove(index)
+    if not index then return false end
+
+    -- Pop the table elements
+    for i=index, #self._ratios do
+        self._ratios[i] = self._ratios[i+1]
+    end
+
+    local ret = fixed.remove(self, index)
+
+    normalize(self)
+
+    return ret
+end
+
+--- Insert a new widget in the layout at position `index`
+-- @param index The position
+-- @param widget The widget
+function ratio:insert(index, widget)
+
+    fixed.insert(self, index, widget)
+
+    normalize(self)
```

Comment:
Yup, please do not base this on `fixed`. That just makes it a lot harder to understand, but doesn't make the code that much shorter.